pkg/sql/tenant_update.go

```diff
@@ -32,45 +33,91 @@ import (
 //
 // Caller is expected to check the user's permission.
 func UpdateTenantRecord(
-	ctx context.Context, settings *cluster.Settings, txn isql.Txn, info *descpb.TenantInfo,
+	ctx context.Context, settings *cluster.Settings, txn isql.Txn, info *mtinfopb.ExtendedTenantInfo,
 ) error {
-	if err := validateTenantInfo(info); err != nil {
+	if err := validateTenantInfo(ctx, settings, info); err != nil {
 		return err
 	}
 
-	tenID := info.ID
-	active := info.State == descpb.TenantInfo_ACTIVE
-	infoBytes, err := protoutil.Marshal(info)
+	// Populate the deprecated DataState field for compatibility
+	// with pre-v23.1 servers.
+	switch info.DataState {
+	case mtinfopb.DataStateReady:
+		info.DeprecatedDataState = mtinfopb.TenantInfo_READY
+	case mtinfopb.DataStateAdd:
+		info.DeprecatedDataState = mtinfopb.TenantInfo_ADD
+	case mtinfopb.DataStateDrop:
+		info.DeprecatedDataState = mtinfopb.TenantInfo_DROP
+	default:
+		return errors.AssertionFailedf("unhandled: %d", info.DataState)
+	}
+	// For the benefit of pre-v23.1 servers.
+	info.DeprecatedID = info.ID
+
+	infoBytes, err := protoutil.Marshal(&info.TenantInfo)
 	if err != nil {
 		return err
 	}
+	// active is a deprecated column preserved for compatibiliy
+	// with pre-v23.1.
+	active := info.DataState == mtinfopb.DataStateReady
+	var name tree.Datum
+	if info.Name != "" {
+		name = tree.NewDString(string(info.Name))
+	} else {
+		name = tree.DNull
+	}
+
+	query := `UPDATE system.tenants
+SET active = $2, info = $3, name = $4, data_state = $5, service_mode = $6
+WHERE id = $1`
+	args := []interface{}{info.ID, active, infoBytes, name, info.DataState, info.ServiceMode}
+	if !settings.Version.IsActive(ctx, clusterversion.V23_1TenantNamesStateAndServiceMode) {
```

Comment:
this does mostly works, but technically is not bulletproof. The better thing to do would be to resolve the table descriptor and see if it has the new columns in a writable state. I won't block this PR over that though. To do that, consider making the `isql.Txn` into a `descs.Txn` in order to access the descriptors.